File: docs/docs/API-Reference/javascript-examples/workflows-api/example-stream-agui-parse.js
const base = (process.env.LANGFLOW_URL ?? process.env.LANGFLOW_SERVER_URL ?? "").replace(/\/$/, "");
const flowId = process.env.FLOW_ID ?? "67ccd2be-17f0-8190-81ff-3bb2cf6508e6";
const apiKey = process.env.LANGFLOW_API_KEY ?? "";
const sessionId = process.env.AGUI_SESSION_ID ?? "thread-123";

const NUMBER = /-?\d+(?:\.\d+)?/;

function numbersFromToolResult(raw) {
  if (raw == null) {
    return [];
  }
  if (typeof raw === "number") {
    return [String(raw)];
  }
  if (typeof raw === "string") {
    return raw.match(new RegExp(NUMBER.source, "g")) ?? [];
  }
  if (typeof raw === "object") {
    for (const key of ["content", "result", "output", "text"]) {
      const value = raw[key];
      if (value == null) {
        continue;
      }
      if (typeof value === "number") {
        return [String(value)];
      }
      const asText = typeof value === "string" ? value : JSON.stringify(value);
      const matches = asText.match(new RegExp(NUMBER.source, "g"));
      if (matches?.length) {
        return matches;
      }
    }
    const serialized = JSON.stringify(raw);
    return serialized.match(new RegExp(NUMBER.source, "g")) ?? [];
  }
  return String(raw).match(new RegExp(NUMBER.source, "g")) ?? [];
}

function extractNumber(text, toolResults) {
  for (let i = toolResults.length - 1; i >= 0; i -= 1) {
    const matches = numbersFromToolResult(toolResults[i]);
    if (matches.length) {
      return Number(matches[matches.length - 1]);
    }
  }
  const matches = text.match(new RegExp(NUMBER.source, "g"));
  return matches?.length ? Number(matches[matches.length - 1]) : null;
}

async function ask(prompt) {
  let text = "";
  const toolResults = [];

  const response = await fetch(`${base}/api/v2/workflows`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": apiKey,
      Accept: "text/event-stream",
    },
    body: JSON.stringify({
      flow_id: flowId,
      input_value: prompt,
      mode: "stream",
      stream_protocol: "agui",
      session_id: sessionId,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${await response.text()}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";
    for (const line of lines) {
      if (!line.startsWith("data:")) {
        continue;
      }
      const event = JSON.parse(line.slice(5).trim());
      switch (event.type) {
        case "TEXT_MESSAGE_CONTENT":
          text += event.delta ?? "";
          break;
        case "TOOL_CALL_RESULT":
          toolResults.push(event.content ?? event.result);
          break;
        case "RUN_ERROR":
          throw new Error(event.message ?? "Run failed");
        case "RUN_FINISHED":
          await reader.cancel();
          return { text, toolResults };
        default:
          break;
      }
    }
  }

  return { text, toolResults };
}

async function main() {
  const prompt1 = process.env.AGUI_PROMPT1 ?? "What is 847 divided by 7?";
  const multiplier = Number(process.env.AGUI_MULTIPLIER ?? "3");

  console.log(`User: ${prompt1}`);
  const { text: reply1, toolResults: tools1 } = await ask(prompt1);
  const quotient = extractNumber(reply1, tools1);
  if (quotient == null) {
    throw new Error("Could not extract a number from run 1.");
  }
  console.log(`Assistant: ${reply1.trim()}`);
  console.log(`Extracted: ${quotient}`);

  const prompt2 = process.env.AGUI_PROMPT2 ?? `Now multiply ${quotient} by ${multiplier}.`;
  console.log(`\nUser: ${prompt2}`);
  const { text: reply2, toolResults: tools2 } = await ask(prompt2);
  const product = extractNumber(reply2, tools2);
  if (product == null) {
    throw new Error("Could not extract a number from run 2.");
  }
  console.log(`Assistant: ${reply2.trim()}`);

  console.log("\n=== Calculation chain ===");
  console.log(`847 ÷ 7 = ${quotient}`);
  console.log(`${quotient} × ${multiplier} = ${product}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: docs/docs/API-Reference/python-examples/workflows-api/example-stream-agui-parse.py
import json
import os
import re

import requests

base = os.environ.get("LANGFLOW_URL") or os.environ.get("LANGFLOW_SERVER_URL", "")
flow_id = os.environ.get("FLOW_ID", "")
api_key = os.environ.get("LANGFLOW_API_KEY", "")
session_id = os.environ.get("AGUI_SESSION_ID", "thread-123")
number = re.compile(r"-?\d+(?:\.\d+)?")


def extract_number(text: str, tool_results: list) -> float | None:
    """Prefer calculator tool output, then the last number in the assistant reply."""
    for raw in reversed(tool_results):
        if matches := number.findall(str(raw)):
            return float(matches[-1])
    if matches := number.findall(text):
        return float(matches[-1])
    return None


def ask(prompt: str) -> tuple[str, list]:
    """Stream one AG-UI run; return assistant text and tool results."""
    text = ""
    tool_results = []

    with requests.post(
        f"{base}/api/v2/workflows",
        headers={
            "Content-Type": "application/json",
            "x-api-key": api_key,
            "Accept": "text/event-stream",
        },
        json={
            "flow_id": flow_id,
            "input_value": prompt,
            "mode": "stream",
            "stream_protocol": "agui",
            "session_id": session_id,
        },
        stream=True,
        timeout=120,
    ) as response:
        response.raise_for_status()
        for line in response.iter_lines(decode_unicode=True):
            if not line or not line.startswith("data:"):
                continue
            event = json.loads(line.removeprefix("data:").strip())
            match event.get("type"):
                case "TEXT_MESSAGE_CONTENT":
                    text += event.get("delta", "")
                case "TOOL_CALL_RESULT":
                    tool_results.append(event.get("content") or event.get("result"))
                case "RUN_ERROR":
                    raise RuntimeError(event.get("message") or "Run failed")
                case "RUN_FINISHED":
                    break

    return text, tool_results


prompt1 = os.environ.get("AGUI_PROMPT1", "What is 847 divided by 7?")
multiplier = int(os.environ.get("AGUI_MULTIPLIER", "3"))

print(f"User: {prompt1}")
reply1, tools1 = ask(prompt1)
quotient = extract_number(reply1, tools1)
if quotient is None:
    raise SystemExit("Could not extract a number from run 1.")
print(f"Assistant: {reply1.strip()}")
print(f"Extracted: {quotient:g}")

prompt2 = os.environ.get("AGUI_PROMPT2") or f"Now multiply {quotient:g} by {multiplier}."
print(f"\nUser: {prompt2}")
reply2, tools2 = ask(prompt2)
product = extract_number(reply2, tools2)
if product is None:
    raise SystemExit("Could not extract a number from run 2.")
print(f"Assistant: {reply2.strip()}")

print("\n=== Calculation chain ===")
print(f"847 ÷ 7 = {quotient:g}")
print(f"{quotient:g} × {multiplier} = {product:g}")
